### PR TITLE
run module uninstall.sh on Magisk uninstallation

### PR DESCRIPTION
**Pull request by HuskyDG**
Currently Magisk does not run `uninstall.sh` script of modules when running Magisk's `uninstaller.sh`

### CODE REVIEW

**Comment by Displax** [comment]
Confirm the problem

**Comment by alnabeel11** [comment]
شكرا جزيلا على المساعدة

**Review by osm0sis — CHANGES_REQUESTED** [review]
"Running module uninstallers" would better match the tense and grammar of the rest of the uninstaller output.

Good fix!

**Review by osm0sis — CHANGES_REQUESTED** [review]
Output should still be plural as "uninstallers" to emphasize it's doing it for any/all modules. Thanks!

**Review by yujincheng08 — CHANGES_REQUESTED** [review]
Why not directly `magisk --remove-modules`

**Comment by yujincheng08** [comment]
I don't think you should run third-party code during recovery (as it is now). Consider a module that wants to delete something in /data during uninstallation, but the /data is encrypted. You can check `BOOTMODE`, and only run `magisk --remove-modules` if it's not in the recovery.

**Comment by yujincheng08** [comment]
Maybe we can add an option to disable the reboot.

File: native/src/core/magisk.cpp [hunk 1]
```diff
@@ -113,8 +114,13 @@ int magisk_main(int argc, char *argv[]) {
                 return 0;
             printf("%s\n", res.data());
         }
+    } else if (argc > 2 && argv[1] == "--remove-modules"sv && argv[2] == "-n"sv) {
+        int fd = connect_daemon(MainRequest::REMOVE_MODULES);
+        write_int(fd, 0);
+        return read_int(fd);
     } else if (argv[1] == "--remove-modules"sv) {
         int fd = connect_daemon(MainRequest::REMOVE_MODULES);
+        write_int(fd, 1);
```
Comment (yujincheng08) [inline native/src/core/magisk.cpp:123]:
```suggestion
        if (argc == 3 && argv[2] == "-n"sv) {
            write_int(fd, 0);
        } else if (argc == 2) {
            write_int(fd, 1);
        } else {
            usage();
            exit(1);
        }
```

File: native/src/core/magisk.cpp [hunk 2]
```diff
@@ -113,8 +114,13 @@ int magisk_main(int argc, char *argv[]) {
                 return 0;
             printf("%s\n", res.data());
         }
+    } else if (argc > 2 && argv[1] == "--remove-modules"sv && argv[2] == "-n"sv) {
```
Comment (yujincheng08) [inline native/src/core/magisk.cpp:117]:
```suggestion
```

File: native/src/core/magisk.cpp [hunk 3]
```diff
@@ -24,6 +24,7 @@ Usage: magisk [applet [arguments]...]
    -V                        print running daemon version code
    --list                    list all available applets
    --remove-modules          remove all modules and reboot
+   --remove-modules -n       remove all modules without reboot
```
Comment (yujincheng08) [inline native/src/core/magisk.cpp:27]:
```suggestion
   --remove-modules [-n]     remove all modules and reboot if -n is not given
```

File: native/src/core/magisk.cpp [hunk 4]
```diff
@@ -113,8 +114,13 @@ int magisk_main(int argc, char *argv[]) {
                 return 0;
             printf("%s\n", res.data());
         }
+    } else if (argc > 2 && argv[1] == "--remove-modules"sv && argv[2] == "-n"sv) {
+        int fd = connect_daemon(MainRequest::REMOVE_MODULES);
+        write_int(fd, 0);
+        return read_int(fd);
```
Comment (yujincheng08) [inline native/src/core/magisk.cpp:120]:
```suggestion
```

File: native/src/core/daemon.cpp [hunk 5]
```diff
@@ -159,10 +159,11 @@ static void handle_request_async(int client, int code, const sock_cred &cred) {
         exec_sql(client);
         break;
     case MainRequest::REMOVE_MODULES:
+        int do_reboot = read_int(client);
         remove_modules();
         write_int(client, 0);
         close(client);
-        reboot();
+        if (do_reboot) reboot();
         break;
```
Comment (yujincheng08) [inline native/src/core/daemon.cpp:167]:
```suggestion
        break;
    }
```

File: native/src/core/daemon.cpp [hunk 6]
```diff
@@ -159,10 +159,11 @@ static void handle_request_async(int client, int code, const sock_cred &cred) {
         exec_sql(client);
         break;
     case MainRequest::REMOVE_MODULES:
```
Comment (yujincheng08) [inline native/src/core/daemon.cpp:161]:
```suggestion
    case MainRequest::REMOVE_MODULES: {
```

File: native/src/core/magisk.cpp [hunk 7]
```diff
@@ -115,6 +115,14 @@ int magisk_main(int argc, char *argv[]) {
         }
     } else if (argv[1] == "--remove-modules"sv) {
         int fd = connect_daemon(MainRequest::REMOVE_MODULES);
```
Comment (yujincheng08) [inline native/src/core/magisk.cpp:117]: we should connect to daemon after checking the parameters

File: native/src/core/daemon.cpp [hunk 8]
```diff
@@ -136,6 +136,7 @@ static void poll_ctrl_handler(pollfd *pfd) {
 }
 
 static void handle_request_async(int client, int code, const sock_cred &cred) {
+    int do_reboot;
```
Comment (yujincheng08) [inline native/src/core/daemon.cpp:139]: You dont need to put it here!

File: native/src/core/magisk.cpp [hunk 9]
```diff
@@ -114,7 +114,17 @@ int magisk_main(int argc, char *argv[]) {
             printf("%s\n", res.data());
         }
     } else if (argv[1] == "--remove-modules"sv) {
+        int do_reboot;
+        if (argc == 3 && argv[2] == "-n"sv) {
+            do_reboot=0;
```
Comment (yujincheng08) [inline native/src/core/magisk.cpp:119]: format the code!!

File: scripts/uninstaller.sh [hunk 10]
```diff
@@ -137,6 +137,11 @@ case $((STATUS & 3)) in
     ;;
 esac
 
+ui_print "- Running module uninstallers"
```
Comment (yujincheng08) [inline scripts/uninstaller.sh:140]: Place it into the branch

**Review by yujincheng08 — APPROVED** [review]

**Comment by osm0sis** [comment]
Considering the uninstaller does setup_flashable and mount_partitions it should be fine also having the originally PR'd script as an else for recovery running of the Magisk uninstaller, no? 🤔

**Review by osm0sis — APPROVED** [review]

**Comment by alnabeel11** [comment]
شكرا جزيلا على المساعدة


في الثلاثاء، ٢٧ ديسمبر ٢٠٢٢ ٧:١٧ ص Chris Renshaw ***@***.***>
كتب:

**Review by topjohnwu — APPROVED** [review]

**Pull request merged**